File: frontend/src/services/adminService.js
// src/services/adminService.js
import api from './api'

export const adminService = {

  // ─── User Management ──────────────────────────────────────────────────────

  getAllUsers: async (params = {}) => {
    try {
      console.log('📡 Fetching users with params:', params)
      const response = await api.get('/admin/users', { params })
      return response.data
    } catch (error) {
      console.error('❌ Get Users Error:', error)
      throw new Error(error.response?.data?.message || 'Failed to fetch users')
    }
  },

  getUserById: async (id) => {
    try {
      const response = await api.get(`/admin/users/${id}`)
      return response.data
    } catch (error) {
      throw new Error(error.response?.data?.message || 'Failed to fetch user')
    }
  },

  changeUserRole: async (userId, role) => {
    try {
      console.log(`🔄 Changing user ${userId} role to ${role}`)
      const response = await api.put(`/admin/users/${userId}/role`, { role })
      return response.data
    } catch (error) {
      throw new Error(error.response?.data?.message || 'Failed to change role')
    }
  },

  toggleUserBan: async (userId) => {
    try {
      console.log(`🚫 Toggling ban for user ${userId}`)
      const response = await api.put(`/admin/users/${userId}/ban`)
      return response.data
    } catch (error) {
      throw new Error(error.response?.data?.message || 'Failed to toggle ban')
    }
  },

  deleteUser: async (userId) => {
    try {
      console.log(`🗑️ Deleting user ${userId}`)
      const response = await api.delete(`/admin/users/${userId}`)
      return response.data
    } catch (error) {
      throw new Error(error.response?.data?.message || 'Failed to delete user')
    }
  },

  getStatistics: async () => {
    try {
      console.log('📊 Fetching statistics')
      const response = await api.get('/admin/stats')
      return response.data
    } catch (error) {
      throw new Error(error.response?.data?.message || 'Failed to fetch statistics')
    }
  },

  getRecentUsers: async (limit = 10) => {
    try {
      const response = await api.get('/admin/users/recent', { params: { limit } })
      return response.data
    } catch (error) {
      throw new Error(error.response?.data?.message || 'Failed to fetch recent users')
    }
  },

  getUsersByRole: async () => {
    try {
      const response = await api.get('/admin/users/by-role')
      return response.data
    } catch (error) {
      throw new Error(error.response?.data?.message || 'Failed to fetch user counts')
    }
  },

  // ─── Vet Clinic Management (NEW) ─────────────────────────────────────────

  getAllClinics: async () => {
    try {
      console.log('🏥 Fetching all vet clinics')
      const response = await api.get('/vets')
      return response.data
    } catch (error) {
      console.error('❌ Get Clinics Error:', error)
      throw new Error(error.response?.data?.message || 'Failed to fetch clinics')
    }
  },

  getClinicById: async (id) => {
    try {
      const response = await api.get(`/vets/${id}`)
      return response.data
    } catch (error) {
      throw new Error(error.response?.data?.message || 'Failed to fetch clinic')
    }
  },

  createClinic: async (clinicData) => {
    try {
      console.log('➕ Creating vet clinic:', clinicData.name)
      const response = await api.post('/vets', clinicData)
      console.log('✅ Clinic created:', response.data)
      return response.data
    } catch (error) {
      console.error('❌ Create Clinic Error:', error)
      throw new Error(error.response?.data?.message || 'Failed to create clinic')
    }
  },

  updateClinic: async (id, clinicData) => {
    try {
      console.log(`✏️ Updating vet clinic ${id}`)
      const response = await api.put(`/vets/${id}`, clinicData)
      console.log('✅ Clinic updated:', response.data)
      return response.data
    } catch (error) {
      console.error('❌ Update Clinic Error:', error)
      throw new Error(error.response?.data?.message || 'Failed to update clinic')
    }
  },

  deleteClinic: async (id) => {
    try {
      console.log(`🗑️ Deleting vet clinic ${id}`)
      const response = await api.delete(`/vets/${id}`)
      console.log('✅ Clinic deleted')
      return response.data
    } catch (error) {
      console.error('❌ Delete Clinic Error:', error)
      throw new Error(error.response?.data?.message || 'Failed to delete clinic')
    }
  },
}

export default adminService

File: backend/src/main/java/backend/admin/service/AdminService.java
package backend.admin.service;

import backend.common.dto.PagedResponse;
import backend.common.exception.BadRequestException;
import backend.common.exception.ResourceNotFoundException;
import backend.user.dto.UserDTO;
import backend.user.model.Role;
import backend.user.model.User;
import backend.user.repository.UserRepository;
import backend.admin.dto.UserStatsDTO;
import backend.pet.repository.PetRepository;
import backend.community.repository.PostRepository;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.data.domain.*;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.time.LocalDateTime;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.stream.Collectors;

@Service
@RequiredArgsConstructor
@Slf4j
public class AdminService {

    private final UserRepository userRepository;
    private final PetRepository petRepository;
    private final PostRepository postRepository;

    public PagedResponse<UserDTO> getAllUsers(int page, int size, Role role, String search, String sortBy, String sortOrder) {
        log.info("Admin fetching users - page: {}, size: {}, role: {}, search: {}", page, size, role, search);

        Sort.Direction direction = sortOrder.equalsIgnoreCase("asc") ? Sort.Direction.ASC : Sort.Direction.DESC;
        Pageable pageable = PageRequest.of(page, size, Sort.by(direction, sortBy));

        Page<User> users;

        if (search != null && !search.trim().isEmpty()) {
            // Search by name or email
            users = userRepository.findByFirstNameContainingIgnoreCaseOrLastNameContainingIgnoreCaseOrEmailContainingIgnoreCase(
                    search, search, search, pageable);
        } else if (role != null) {
            // Filter by role
            users = userRepository.findByRole(role, pageable);
        } else {
            // Get all users
            users = userRepository.findAll(pageable);
        }

        List<UserDTO> userDTOs = users.getContent().stream()
                .map(this::mapToDTO)
                .collect(Collectors.toList());

        return new PagedResponse<>(
                userDTOs,
                users.getNumber(),
                users.getSize(),
                users.getTotalElements(),
                users.getTotalPages(),
                users.isLast()
        );
    }

    public UserDTO getUserById(Long id) {
        User user = userRepository.findById(id)
                .orElseThrow(() -> new ResourceNotFoundException("User not found"));
        return mapToDTO(user);

    }

    @Transactional
    public UserDTO changeUserRole(Long id, Role newRole) {
        log.info("Changing user {} role to {}", id, newRole);

        User user = userRepository.findById(id)
                .orElseThrow(() -> new ResourceNotFoundException("User not found"));

        user.setRole(newRole);
        User savedUser = userRepository.save(user);

        log.info("User {} role changed to {} successfully", id, newRole);
        return mapToDTO(savedUser);
    }

    @Transactional
    public UserDTO toggleUserBan(Long id) {
        log.info("Toggling ban status for user {}", id);

        User user = userRepository.findById(id)
                .orElseThrow(() -> new ResourceNotFoundException("User not found"));

        user.setEnabled(!user.isEnabled());
        User savedUser = userRepository.save(user);

        log.info("User {} ban status toggled. Enabled: {}", id, savedUser.isEnabled());
        return mapToDTO(savedUser);
    }

    @Transactional
    public void deleteUser(Long id) {
        log.info("Deleting user {}", id);

        User user = userRepository.findById(id)
                .orElseThrow(() -> new ResourceNotFoundException("User not found"));

        // Prevent deleting yourself
        // You might want to add current user check here

        userRepository.delete(user);
        log.info("User {} deleted successfully", id);
    }

    public UserStatsDTO getStatistics() {
        long totalUsers = userRepository.count();
        long ownersCount = userRepository.countByRole(Role.OWNER);
        long vetsCount = userRepository.countByRole(Role.VET);
        long adminsCount = userRepository.countByRole(Role.ADMIN);
        long totalPets = petRepository.count();
        long totalPosts = postRepository.count();

        // Get users registered in last 30 days
        LocalDateTime thirtyDaysAgo = LocalDateTime.now().minusDays(30);
        long newUsersLast30Days = userRepository.countByCreatedAtAfter(thirtyDaysAgo);

        return UserStatsDTO.builder()
                .totalUsers(totalUsers)
                .ownersCount(ownersCount)
                .vetsCount(vetsCount)
                .adminsCount(adminsCount)
                .totalPets(totalPets)
                .totalPosts(totalPosts)
                .newUsersLast30Days(newUsersLast30Days)
                .build();
    }

    public List<UserDTO> getRecentUsers(int limit) {
        Pageable pageable = PageRequest.of(0, limit, Sort.by(Sort.Direction.DESC, "createdAt"));
        return userRepository.findAll(pageable).getContent().stream()
                .map(this::mapToDTO)
                .collect(Collectors.toList());
    }

    public Map<String, Long> getUserCountByRole() {
        Map<String, Long> roleCounts = new HashMap<>();
        roleCounts.put("OWNER", userRepository.countByRole(Role.OWNER));
        roleCounts.put("VET", userRepository.countByRole(Role.VET));
        roleCounts.put("ADMIN", userRepository.countByRole(Role.ADMIN));
        return roleCounts;
    }

    private UserDTO mapToDTO(User user) {
        return UserDTO.builder()
                .id(user.getId())
                .firstName(user.getFirstName())
                .lastName(user.getLastName())
                .name(user.getFullName())
                .email(user.getEmail())
                .phone(user.getPhone())
                .address(user.getAddress())
                .avatar(user.getAvatar())
                .role(user.getRole())
                .enabled(user.isEnabled())
                .createdAt(user.getCreatedAt())
                .build();
    }
}